src/LinAlg/hiopVectorCuda.cpp

```diff
@@ -0,0 +1,1118 @@
+  return nrm1_global;
+#endif
+  return norm1;
+}
+
+double hiopVectorCuda::onenorm_local() const
+{
+//  double* data = data_dev_;
+//  int n = n_local_;
+//  return hiop::cuda::onenorm_local_kernel(n, data);
+    return hiop::cuda::onenorm_local_kernel(n_local_, data_dev_);
+}
+
+void hiopVectorCuda::componentMult( const hiopVector& vec )
+{
+  const hiopVectorCuda& v = dynamic_cast<const hiopVectorCuda&>(vec);
+  assert(n_local_ == v.n_local_);
+
+  hiop::cuda::thrust_component_mult_kernel(n_local_, data_dev_, v.data_dev_);
+}
+
+void hiopVectorCuda::componentDiv( const hiopVector& vec )
+{
+  const hiopVectorCuda& v = dynamic_cast<const hiopVectorCuda&>(vec);
+  assert(n_local_ == v.n_local_);
+
+  hiop::cuda::thrust_component_div_kernel(n_local_, data_dev_, v.data_dev_);
+}
+
+void hiopVectorCuda::componentDiv_w_selectPattern( const hiopVector& vec, const hiopVector& select)
+{
+  const hiopVectorCuda& v = dynamic_cast<const hiopVectorCuda&>(vec);
+  assert(n_local_ == v.n_local_);
+
+  const hiopVectorCuda& ix = dynamic_cast<const hiopVectorCuda&>(select);
+  hiop::cuda::component_div_w_pattern_kernel(n_local_, data_dev_, v.data_dev_, ix.data_dev_);
+}
+
+void hiopVectorCuda::component_min(const double constant)
+{
+  hiop::cuda::component_min_kernel(n_local_, data_dev_, constant);
+}
+
+void hiopVectorCuda::component_min(const hiopVector& vec)
+{
+  const hiopVectorCuda& v = dynamic_cast<const hiopVectorCuda&>(vec); 
+  assert(v.get_local_size() == n_local_);
+
+  const double* vd = v.data_dev_;
+  
+  hiop::cuda::component_min_kernel(n_local_, data_dev_, vd);
+}
+
+void hiopVectorCuda::component_max(const double constant)
+{
+  hiop::cuda::component_max_kernel(n_local_, data_dev_, constant);
+}
+
+void hiopVectorCuda::component_max(const hiopVector& vec)
+{
+  const hiopVectorCuda& v = dynamic_cast<const hiopVectorCuda&>(vec); 
+  assert(v.get_local_size() == n_local_);
+
+  const double* vd = v.data_dev_;
+  
+  hiop::cuda::component_max_kernel(n_local_, data_dev_, vd);
+}
+
+void hiopVectorCuda::component_abs()
+{
+  hiop::cuda::thrust_component_abs_kernel(n_local_, data_dev_);
+}
+
+void hiopVectorCuda::component_sgn ()
+{
+  hiop::cuda::thrust_component_sgn_kernel(n_local_, data_dev_);
+}
+
+void hiopVectorCuda::component_sqrt()
+{
+  hiop::cuda::thrust_component_sqrt_kernel(n_local_, data_dev_);
+}
+
+/// Scale each element of this  by the constant alpha
+void hiopVectorCuda::scale(double alpha)
+{
+  int one = 1;  
+  cublasStatus_t ret_cublas = cublasDscal(handle_cublas_, n_local_, &alpha, data_dev_, one);
+  assert(ret_cublas == CUBLAS_STATUS_SUCCESS);
+}
+
+/// Implementation of AXPY kernel
+void hiopVectorCuda::axpy(double alpha, const hiopVector& xvec)
+{
+  const hiopVectorCuda& x = dynamic_cast<const hiopVectorCuda&>(xvec);
+  int one = 1;
+  cublasStatus_t ret_cublas = cublasDaxpy(handle_cublas_, n_local_, &alpha, x.data_dev_, one, data_dev_, one);
+  assert(ret_cublas == CUBLAS_STATUS_SUCCESS);
+}
```

Comment:
I think if you use abstract interface method to access local vector data, you can avoid dynamic casting. Perhaps if local data accessor is a protected method in the base class that would be a good solution here?
```suggestion
void hiopVectorCuda::axpy(double alpha, const hiopVector& xvec)
{
  int one = 1;
  cublasStatus_t ret_cublas = cublasDaxpy(handle_cublas_, n_local_, &alpha, x.data(), one, data_dev_, one);
  assert(ret_cublas == CUBLAS_STATUS_SUCCESS);
}
```

Reply:
Same at several other places in the code.